### ISSUE TITLE
`ImmersedBoundaryCondition` has the opposite convention of `FieldBoundaryConditions`

### ISSUE THREAD

**Issue by simone-silvestri** [CliMA/Oceananigans.jl#3141]
The current implementation of `ImmersedBondaryCondition` has opposite conventions as `FieldBoundaryConditions`.

for example, to have a bottom drag, you have to write
```
u_bc = FieldBoundaryConditions(bottom = drag_u)
```
but to have a drag above an immersed boundary
```
u_immersed = ImmersedBoundaryCondition(top = drag_u)
```

I think it is better to assume that immersed and non-immersed boundaries have the same conventions. (I ran for a lot of time an ocean simulation without drag because I haven't looked at the implementation and assumed the convention was the same)

I propose to change the convention to have 
```
u_immersed = ImmersedBoundaryCondition(bottom = drag_u)
```

(I guess we will need to invert `index_left` and `index_right` in 
https://github.com/CliMA/Oceananigans.jl/blob/6ab0c77ec7cb51b768ee30e248d633c98bb7e427/src/ImmersedBoundaries/immersed_boundary_condition.jl#L200-L201)

found by @Yixiao-Zhang

**Comment by simone-silvestri** [comment]
Actually looking closely about it, it looks like a bug, because in `divergence_operators.jl` the indices are defined as I would expect

https://github.com/CliMA/Oceananigans.jl/blob/6ab0c77ec7cb51b768ee30e248d633c98bb7e427/src/Operators/divergence_operators.jl#L78

**Comment by glwagner** [comment]
There is a problem with our convention for fluxes as we have discussed before.

We can change the convention for fluxes on non-immersed boundaries to fix it. The usual convention is to say that fluxes point inwards, into the domain. This would change the sign of surface fluxes.

For users who specify fluxes on individual facets of immersed boundary cells this is less of an issue. However, if you wish to specify a flux across immersed boundaries in general, you have to use a convention like “fluxes point on the inward normal”.

I’m not sure about a separate bug, though.

**Comment by glwagner** [comment]
So to summarize: right now the convention for ImmersedBoundaryCondition is that fluxes orient along the inward normal. This means that for a “top” boundary condition, the convention is opposite what we would use at the top of the domain (usually the ocean surface).

i think the way out of this puzzle is to adopt the same convention for fluxes on all boundaries. But it’s a huge breaking change and we probably need to put a warning in the code that will persist for a while…

**Comment by glwagner** [comment]
Ok, I agree that is a bug if that's how it is implemented.

I was responding to this:

> The current implementation of ImmersedBondaryCondition has opposite conventions as FieldBoundaryConditions.

which is also true.

**Comment by glwagner** [comment]
If the issue is just the names for the facets of an immersed cell, then it's an easy fix since you just have to relabel the boundaries without changing any implementation.

**Comment by simone-silvestri** [comment]
:+1: I ll take a look then and switch around the labels and make sure that it works.

**Comment by glwagner** [comment]
Note that there is a validation test for couette flow here:

https://github.com/CliMA/Oceananigans.jl/blob/main/validation/immersed_boundaries/immersed_couette_flow.jl

So apparently this is only a problem with the API for specifying fluxes on individual boundary facets. Ie this works:

```julia
u_immersed_bc = ValueBoundaryCondition(-1)
u_top_bc = ValueBoundaryCondition(1)
u_bcs = FieldBoundaryConditions(immersed=u_immersed_bc, top=u_top_bc)
```

**Comment by Yixiao-Zhang** [comment]
Thank you! This solves my problem.

**Comment by glwagner** [comment]
What solves your problem @Yixiao-Zhang ?

**Comment by Yixiao-Zhang** [comment]
I mean the new user interface for `ImmersedBoundaryCondtion`. Previously, I tried to do a simulation with heating distributed on an irregular bottom, and  I found that I had to use `ImmersedBoundaryCondition(top = the_bottom_heating)` instead of `ImmersedBoundaryCondition(bottom = the_bottom_heating)` after some testing. I felt this strange and reported this behavior to Simone. Then he opened this issue, because it affects his simulations as well. I am sorry for confusing you, since you did not know Simone and I had some discussion on issue before.

**Comment by glwagner** [comment]
@Yixiao-Zhang can you set up your simulation without using `ImmersedBoundaryCondition` at all? `ImmersedBoundaryCondition` is not appropriate for most physical modeling. I suggest something like

```julia
FieldBoundaryConditions(immersed=the_bottom_heating)
```

**Comment by Yixiao-Zhang** [comment]
Yes. The changes on #3142 enable me to use `ImmersedBoundaryCondition(bottom = the_bottom_heating)`, which is more natural for me.

Can I ask why `ImmersedBoundaryCondition` is not appropriate for most physical modeling?

**Comment by Yixiao-Zhang** [comment]
Actually, I tried using `FieldBoundaryConditions(immersed=the_bottom_heating)` before I used `ImmersedBoundaryCondition`. The problem is that the total amount of bottom heating would defined in a strange way if I did so. Let's consider a seamount, the height of which is $\eta(x)$. Assume that $F(x)$ is a function for the bottom heat flux. If I use `FieldBoundaryConditions(immersed=F)`, the total bottom heating will be
$$F_\mathrm{total} = \int_{\mathrm{Domain}}  F(x)(1 + |s(x)|) \mathrm{d}x\mathrm{d}y$$
where $s(x)$ is the slope of the seamount:
$$s = \frac{\mathrm{d}\eta}{\mathrm{d}x}$$

This is because fluxes are applies to all facets, and the seamount is represented with "Lego bricks". I have verified this expression with one simulation. However, it has no physical meaning. If $F$ is the flux in the direction of the normal of the surface, the expression for $F_\mathrm{total}$ will be
$$F_\mathrm{total} = \int_{\mathrm{Domain}}  F(x)\sqrt{1 + |s(x)|^2} \mathrm{d}x\mathrm{d}y$$
Or if $F$ is the upward flux,  the expression for $F_\mathrm{total}$ will be
$$F_\mathrm{total} = \int_{\mathrm{Domain}}  F(x)\mathrm{d}x\mathrm{d}y$$

I think the latter two expressions for $F_\mathrm{total}$ are reasonable to me. However, the first one seems too strange to me. It reminds me of the $\pi$-is-4 paradox.

Besides, this is just the case of a one-dimensional seamount. If $\eta$ is a function of $x$ and $y$, the expression for $F_\mathrm{total}$ will be even more complicated when using `FieldBoundaryConditions(immersed=F)`. It will break the rotational symmetry.

**Comment by glwagner** [comment]
Can you elaborate why it is strange? I think the first expression is the correct physical model for heat flux from a bumpy surface. That's why, for example, heat exchangers are designed with a large surface area for heating your home, etc.

**Comment by Yixiao-Zhang** [comment]
I want to simulate convection generated by a given heat flux. The total bottom heating is known. That is why I prefer `ImmersedBoundaryCondition(bottom = the_bottom_heating)`. We know the flux heat out of upward-facing facets from geological constraints.

Can I ask why `ImmersedBoundaryCondition` is not appropriate for most physical modeling? Is it a numerical issue? (I am sorry for asking again, but this seems important.)

**Comment by Yixiao-Zhang** [comment]
I feel the first expression strange because the realistic topography is usually (at least in my case) not Lego-like, even though we use Lego-like topography in numerical simulations. For this reason, the second expression is most natural to me. (However, it seems hard to code.)

**Comment by glwagner** [comment]
Well, ok. I see how using `ImmersedBoundaryCondition` is _convenient_. I disagree that this is a "better" model, however. It is an approximate model that leverages (what is hopefully) the thin aspect ratio of your simulation.

Note, I don't think these issues are intrinsic to lego topography. In fact, I think this approach could completely fail if we had cut cells, which represents another caveat. The problem is that, when the cell facets are no longer oriented along one of x, y, z, we need to define what a flux across the bottom _means_. Is it a flux normal to the boundary, or is it a flux projected along one of the orthogonal directions? I'm not sure interpreting `ImmersedBoundaryCondition(bottom=flux)` as representing _only_ the vertical component of the bottom flux is sensible, especially when we consider the small scale limit or large eddy simulation or direct numerical simulation. We'll have to discuss that.

To elaborate on the details, the geological constraint we have from observations is on the heat flux averaged over large patches of the ocean basin:

$$ Q_\text{obs} \approx \frac{1}{A} \int \boldsymbol{q} \cdot \boldsymbol{n}  d A $$

where $A$ is the surface area of the ocean bottom (not the horizontal area).

As you have noted, we can reproduce this constraint by applying the flux $Q$ across all upward facing facets. Then the total flux in our model is

$$ Q_\text{model} = \frac{1}{A_h} \int Q_\text{obs} d A_h $$

where $A_h$ is the horizontal area of the ocean basin and $d A_h$ is the horizontal projection of the surface area of the bottom. Clearly we obtain $Q_\text{model} = Q_\text{obs}$ and our goal is achieved.

At small scales, this approximate model becomes questionable (but of course, it is also questionable to use a constant heat flux at small scales).

Basically, I think that using this convenience actually obscures the true physics. So while it produces the right answer, it leads to confusion about what the observed geological heat flux actually represents, and the detailed physics of heat flux across irregular boundaries...

That said, I take the very important point that the convenience greatly simplifies applying this kind of model at large scales. So I am more convinced that we should expose `ImmersedBoundaryCondition` to users --- despite the dangers...

**Comment by Yixiao-Zhang** [comment]
I see your point. I agree that using `FieldBoundaryConditions(immersed=the_flux)` generally makes more sense.

In my point of view, the biggest danger is the possible misuse of `ImmersedBoundaryCondition`. When a user uses `ImmersedBoundaryCondition(bottom = the_flux)`, he or she may think that the flux is applied to all facets at the bottom of the domain.

Therefore, I think it would be better if we could rename keywords for `ImmersedBoundaryCondition`. Instead to using which facet the flux is applied to, we can use the direction of the flux. For example, we can use `upward` instead of `bottom`. I think it will make `ImmersedBoundaryCondition` explicit enough that users know what they are doing.

**Comment by glwagner** [comment]
I agree and I think that's also an issue with the "fix" in #3142. Eg previously the kwargs referred to immersed cells, ie `bottom=flux` means the `flux` is applied at the bottom of an immersed cell. But apparently, this was confused with the "bottom" of the domain...

`upward` could be a good alternative. I'm worried about deciding on these conventions before we have cut cells though, which would probably become the preferred default. Whatever API we have should make sense for cut cells. 

All of this is one reason why I feel it might better to remove documentation for `ImmersedBoundaryCondition` altogether and recommend using ordinary boundary conditions on immersed boundaries.

**Comment by glwagner** [comment]
That said I do think `upward`, `eastward`, `westward` could apply to cut cells. So I think that could potentially be a reasonable API to elaborate on. I mostly wonder whether it's premature before we have the cut cell implementation and have had a chance to work with it a bit. @siddharthabishnu

**Comment by Yixiao-Zhang** [comment]
It would probably make sense to have a structure called `ImmersedBoundaryFlux`. If so, we will not have to expose `ImmersedBoundaryCondition`. What matters to me is that the integral of the flux over the ocean bottom (or any surface) should be easy to calculate.

**Comment by Yixiao-Zhang** [comment]
I am a little worried that the possibility of using either Lego topography or cut cells may lead to confusion when calculating the total flux. When switching from Lego topography to cut cells, you may argue that the total flux should decrease since the total surface area becomes smaller. However, in a lot of cases, the total flux is constrained by external processes (like bottom heating constrained by geological processes, for example) rather than the fluid motion near the boundary.

**Comment by glwagner** [comment]
> What matters to me is that the integral of the flux over the ocean bottom (or any surface) should be easy to calculate.

I think that would argue for defining fluxes as normal to boundaries, since you can then use `GridMetricOperation` eg

```julia
using Oceananigans.AbstractOperations: Az
Q_operation = MappedOperation(q * Az, k=BottomCell())
Q = Integral(Q_operation)
```

(once we have the `MappedOperation` feature to do calculations only on bottom cells).
